src/core/apparmor-setup.c

```diff
@@ -0,0 +1,127 @@
+
+#if HAVE_APPARMOR
+DEFINE_TRIVIAL_CLEANUP_FUNC(aa_kernel_interface *, aa_kernel_interface_unref);
+DEFINE_TRIVIAL_CLEANUP_FUNC(aa_features *, aa_features_unref);
+#endif
+
+int mac_apparmor_setup(void) {
+#if HAVE_APPARMOR
+        int r;
+        _cleanup_close_ int fd = -1;
+        _cleanup_(aa_kernel_interface_unrefp) aa_kernel_interface *interface = NULL;
+        _cleanup_(aa_features_unrefp) aa_features *features = NULL;
+        const char *current_file;
+        char **profile_file;
+        _cleanup_strv_free_ char **profile_files = NULL;
+        _cleanup_free_ char *current_profile = NULL, *feature_hash = NULL;
+
+        if (!mac_apparmor_use()) {
+                log_debug("AppArmor either not supported by the kernel or disabled.");
+                return 0;
+        }
+
+        /* To enable LSM stacking a patch to the kernel is proposed to create a
+         * per-LSM subdirectory to distinguish between the LSMs. Therefore, we
+         * read the file from the LSM specific directory first and only if that
+         * fails the one from the generic directory.
+         */
+        FOREACH_STRING(current_file, "/proc/self/attr/apparmor/current", "/proc/self/attr/current") {
+                r = read_one_line_file(current_file, &current_profile);
+                if (r == -ENOENT)
+                        continue;
+                else if (r < 0)
+                        log_warning_errno(r, "Failed to read current AppArmor profile from file %s, ignoring: %m", current_file);
+                else
+                        break;
+        }
+        if (!current_profile) {
+                log_warning("Failed to get the current AppArmor profile of systemd from /proc/self/attr/apparmor/current or /proc/self/attr/current, ignoring.");
+                return 0;
+        }
+        if (!streq(current_profile, "unconfined")) {
+                log_debug("We are already confined in an AppArmor profile.");
+                return 0;
+        }
+
+        r = aa_features_new_from_kernel(&features);
+        if (r < 0) {
+                log_warning_errno(errno, "Failed to get the AppArmor feature set from the kernel, ignoring: %m");
+                return 0;
+        }
+        feature_hash = aa_features_id(features);
+
+        r = strv_extendf(&profile_files, "/var/cache/apparmor/%s/usr.lib.systemd.systemd", feature_hash);
+        if (r < 0)
+                return log_oom();
+        r = strv_extendf(&profile_files, "/etc/systemd/profile.apparmor.%s", feature_hash);
+        if (r < 0)
+                return log_oom();
+        r = strv_extendf(&profile_files, "/usr/lib/systemd/profile.apparmor.%s", feature_hash);
+        if (r < 0)
+                return log_oom();
+        r = strv_extend(&profile_files, "/etc/systemd/profile.apparmor");
+        if (r < 0)
+                return log_oom();
+        r = strv_extend(&profile_files, "/usr/lib/systemd/profile.apparmor");
+        if (r < 0)
+                return log_oom();
```

Comment:
I would drop all of these, in favor of only loading from `/var/cache/apparmor`.

It is the correct and the default location, which all prebuilt / immutable images should have, as well as any "package" (deb, rpm, etc) based systems too.

Why do we still need loading from /etc/systemd, /usr/lib/systemd, if loading from /var/cache is there? Note this is binary profile, rather than the textual representation of the profile.

Otherwise the search paths look fine, and ordering is fine. I just don't think we need them at all.

Reply:
I am ok with that. I will just maintain a small patch privately that adds the other directories back.

Reply:
Sorry, but /var is a no-go for early boot stuff like this, see other comment, sorry.

Reply:
(The other directories make a lot more sense to me than the /var path...)